### PR TITLE
Edge Sync File Id Strategy

### PR DESCRIPTION
**Pull request by kyakdan**

### CODE REVIEW

File: packages/instrumentor/edgeIdStrategy.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+		}
+
+		this._nextEdgeId = this.firstEdgeId;
+	}
+	commitIdCount(filename: string): void {
+		if (this.firstEdgeId === undefined) {
+			throw Error("commitIdCount() is called before startForSourceFile()");
+		}
+
+		const idCount = this._nextEdgeId - this.firstEdgeId;
```
Comment (bertschneider) [inline packages/instrumentor/edgeIdStrategy.ts:161]: The name confused me somehow. Perhaps `usedIdsCount`?

File: packages/core/cli.ts [hunk 2]
```diff
@@ -74,6 +74,15 @@ yargs(process.argv.slice(2))
 				})
 				.hide("fuzzFunction")
 
+				.option("id_sync_file", {
+					describe:
+						"File used to sync edge ID generation. " +
+						"Needed when fuzzing in multi-process modes",
```
Comment (bertschneider) [inline packages/core/cli.ts:80]: Should the fork mode automatically be activated if `id_sync_file` is set? If not we should mention the necessary libFuzzer flags and the default that is used if no sync file is set.

**Comment by bertschneider** [comment]
I reviewed commit by commit and noticed too late, that some comments were on already changed code. Everything should still be valid, I hope, but that's why code on some comments is marked as outdated.

File: packages/instrumentor/edgeIdStrategy.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { fuzzer } from "@jazzer.js/fuzzer";
+
+export abstract class EdgeIdStrategy {
```
[the remark this answers is not in the document]
Comment (kyakdan) [reply packages/instrumentor/edgeIdStrategy.ts]: Added an interface and abstract class that implements the `nextEdgeId` method so that it can be shared by two ID strategies

File: packages/fuzzer/coverage.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+let nextEdgeID = 0;
+
+type NativeAddon = {
+	registerCoverageMap: (buffer: Buffer) => void;
+	registerNewCounters: (oldNumCounters: number, newNumCounters: number) => void;
+};
+
+let nativeAddon: NativeAddon;
+
+export function initializeCounters(addon: NativeAddon) {
```
[the remark this answers is not in the document]
Comment (kyakdan) [reply packages/fuzzer/coverage.ts]: I've extracted coverage and tracing functions into dedicated objects exposed through the `Fuzzer` interface.

File: packages/core/cli.ts (hunk 2 above)
Comment (bertschneider) [reply packages/core/cli.ts:80]: Ah, good point, didn't think about that.

Comment (bertschneider) [reply packages/core/cli.ts:80]: Still, should we mention the affected libFuzzer modes or, if the flag should only be used internally, hide it?

Comment (kyakdan) [reply packages/core/cli.ts:80]: I think hiding the flag would be the better option for now.

Comment (kyakdan) [reply packages/core/cli.ts:80]: Done!

File: packages/instrumentor/edgeIdStrategy.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+				// If the ID sync file is already locked, wait for a random period of time
+				// between 0 and 100 milliseconds. Waiting for different periods reduces
+				// the chance of all processes wanting to acquire the lock at the same time.
+				this.wait(this.randomIntFromInterval(0, 100));
+				continue;
+			}
+			try {
+				// Acquire the lock for the ID sync file and look for the initial edge ID and
+				// corresponding number of inserted counters.
+				lock.lockSync(this.idSyncFile);
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply packages/instrumentor/edgeIdStrategy.ts]: To me, the new version looks simpler. I would leave it that way.

In the new version, we could invoke `releaseLockOnSyncFile`, if set, in the `catch` block, before re-throwing the error. It shouldn't make a huge difference, as the application is going to stop anyway, but the lockfile would be cleaned up in unexpected situations.

File: packages/instrumentor/edgeIdStrategy.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+/**
+ * A strategy for edge ID generation that synchronizes the IDs assigned to a source file
+ * with other processes via the specified `idSyncFile`. The edge information stored as a
+ * line of the format: <source file path>,<initial edge ID>,<total edge count>
+ *
+ * This class takes care of synchronizing the access to the file between
+ * multiple processes accessing it during instrumentation.
+ */
+export class FileSyncIdStrategy extends IncrementingEdgeIdStrategy {
+	private static fatalExitCode = 79;
```
Comment (bertschneider) [inline packages/instrumentor/edgeIdStrategy.ts:73]: `readonly`?

File: packages/instrumentor/edgeIdStrategy.ts [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+/**
+ * A strategy for edge ID generation that synchronizes the IDs assigned to a source file
+ * with other processes via the specified `idSyncFile`. The edge information stored as a
+ * line of the format: <source file path>,<initial edge ID>,<total edge count>
+ *
+ * This class takes care of synchronizing the access to the file between
+ * multiple processes accessing it during instrumentation.
+ */
+export class FileSyncIdStrategy extends IncrementingEdgeIdStrategy {
```
Comment (bertschneider) [inline packages/instrumentor/edgeIdStrategy.ts:72]: We probably should create unit tests for the edge id strategies. If this is time critical, we can do it in a separate PR.

File: packages/instrumentor/instrument.ts [hunk 8]
```diff
@@ -114,16 +135,29 @@ function instrument(
...
 	shouldInstrument: FilePredicate,
-	shouldHook: FilePredicate
+	shouldHook: FilePredicate,
+	idStrategy: EdgeIdStrategy
 ) {
 	const transformations: PluginItem[] = [];
-	if (shouldInstrument(filename)) {
-		transformations.push(codeCoverage, compareHooks);
+	const shouldInstrumentFile = shouldInstrument(filename);
+	if (shouldInstrumentFile) {
```
Comment (bertschneider) [inline packages/instrumentor/instrument.ts:143]: This function gets more and more complex. Nothing to do for now, but we should start thinking about better abstractions to add plugins without needing to extend the function further.

File: packages/instrumentor/plugins/codeCoverage.test.ts [hunk 9]
```diff
@@ -16,13 +16,23 @@
...
+					return 0;
+				},
+			};
+		}),
+	};
+});
 
-const expectInstrumentation = instrumentWith(codeCoverage);
+const expectInstrumentation = instrumentWith(
+	codeCoverage(new MemorySyncIdStrategy())
```
Comment (bertschneider) [inline packages/instrumentor/plugins/codeCoverage.test.ts:34]: We could use a simple implementation of `EdgeIdStrategy`, always returning `0`, for the test. No need for complex mocking anymore.

File: packages/fuzzer/fuzzer.ts [hunk 10]
```diff
@@ -14,191 +14,27 @@
...
-	startFuzzing: (fuzzFn: FuzzTarget, fuzzOpts: FuzzOpts) => void;
-	startFuzzingAsync: (fuzzFn: FuzzTarget, fuzzOpts: FuzzOpts) => Promise<void>;
-	stopFuzzingAsync: (status?: number) => void;
-	nextCounter: typeof nextCounter;
-	incrementCounter: typeof incrementCounter;
-	readCounter: typeof readCounter;
-	traceStrCmp: typeof traceStrCmp;
-	traceNumberCmp: typeof traceNumberCmp;
-	traceAndReturn: typeof traceAndReturn;
+	nativeAddon: NativeAddon;
```
Comment (bertschneider) [inline packages/fuzzer/fuzzer.ts:28]: I really like using properties to encapsulate related functions in `Fuzzer`.

But, IMHO, `NativeAddon` is an implementation detail and shouldn't be exposed outside of the `fuzzer` package. Its function could be divided between `coverageTracer`, `tracer` and another property related to the fuzzer execution / state. WDYT?

File: packages/instrumentor/instrument.ts (hunk 8 above)
Comment (kyakdan) [reply packages/instrumentor/instrument.ts:143]: Agreed. It will become even more complex with the coverage PR.

File: packages/instrumentor/edgeIdStrategy.ts (hunk 6 above)
Comment (kyakdan) [reply packages/instrumentor/edgeIdStrategy.ts:73]: Done!

File: packages/instrumentor/edgeIdStrategy.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kyakdan) [reply packages/instrumentor/edgeIdStrategy.ts]: Done.

File: packages/instrumentor/plugins/codeCoverage.test.ts (hunk 9 above)
Comment (kyakdan) [reply packages/instrumentor/plugins/codeCoverage.test.ts]: Added

File: packages/instrumentor/edgeIdStrategy.ts (hunk 7 above)
Comment (kyakdan) [reply packages/instrumentor/edgeIdStrategy.ts:72]: I've added a unit test for the sync file ID strategy and refactored the interface if the `instrument` functions to make it easier to use and test. Could you have a look?

File: packages/instrumentor/instrument.test.ts [hunk 11]
```diff
@@ -95,8 +98,11 @@
 	const oldFuzzer = globalThis.Fuzzer;
 	// @ts-ignore
 	globalThis.Fuzzer = {
-		incrementCounter: () => {
-			// ignore
+		// @ts-ignore
+		coverageTracker: {
+			incrementCounter: (edgeId: number) => {
```
Comment (bertschneider) [inline packages/instrumentor/instrument.test.ts:103]: Warning that the parameter is not needed, can be removed.

File: packages/instrumentor/instrument.ts [hunk 12]
```diff
@@ -110,20 +122,39 @@ export function shouldInstrumentFn(
...
 	filename: string,
-	shouldInstrument: FilePredicate,
-	shouldHook: FilePredicate
+	options: InstrumentorOptions
 ) {
 	const transformations: PluginItem[] = [];
-	if (shouldInstrument(filename)) {
-		transformations.push(codeCoverage, compareHooks);
+
+	const shouldInstrument = shouldInstrumentFn(
```
Comment (bertschneider) [inline packages/instrumentor/instrument.ts:132]: `shouldInstrument` and `shouldHook` are created for every file. If they should stay here, we could remove one level of indirection and directly call the functions with all needed arguments and so get rid of the higher-order function `shouldInstrumentFn` and `bind` on `hasFunctionsToHook`. Another option would be to pass in the functions in the `InstrumentorOptions`. I think I would like that more.

File: packages/instrumentor/instrument.ts [hunk 13]
```diff
@@ -110,20 +122,39 @@ export function shouldInstrumentFn(
 	};
 }
 
-function instrument(
+export function instrument(
 	code: string,
 	filename: string,
-	shouldInstrument: FilePredicate,
-	shouldHook: FilePredicate
+	options: InstrumentorOptions
```
Comment (bertschneider) [inline packages/instrumentor/instrument.ts:128]: Would destructuring `options` make the function body simpler?

File: packages/instrumentor/instrument.test.ts (hunk 11 above)
Comment (kyakdan) [reply packages/instrumentor/instrument.test.ts:103]: Removed.

Comment (kyakdan) [reply packages/instrumentor/instrument.test.ts:102]: The could will not compile if we remove this ignore directive

File: packages/instrumentor/instrument.ts (hunk 12 above)
Comment (kyakdan) [reply packages/instrumentor/instrument.ts:132]: As discussed, I've refactored the instrumentor code into a dedicated class.

Comment (bertschneider) [reply packages/instrumentor/instrument.ts]: :+1:
